File: backend/app/api/v1/follow.py
"""
关注接口 - 关注/取关、粉丝列表、关注列表；与技术方案 3.3.2 一致
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user
from app.core.exceptions import NotFoundException, AuthorizationException, success_response
from app.models.user import User
from app.schemas.follow import FollowActionResponse, FollowListResponse
from app.schemas.user import UserResponse
from app.schemas.post import PostResponse
from app.services.follow_service import (
    count_following_posts,
    follow_user,
    unfollow_user,
    get_followers,
    get_following,
    is_following,
    list_following_posts,
)
from app.services.user_service import get_user_by_id

router = APIRouter(prefix="/user", tags=["follow"])


def _user_to_response(user: User) -> dict:
    return {
        "id": user.id,
        "anonymous_name": user.anonymous_name,
        "avatar": user.avatar,
        "bio": user.bio,
        "follower_count": user.follower_count or 0,
        "following_count": user.following_count or 0,
        "post_count": user.post_count or 0,
        "allow_follow": user.allow_follow,
        "allow_comment": user.allow_comment,
        "status": user.status,
        "created_at": user.created_at.isoformat() if user.created_at else None,
        "updated_at": user.updated_at.isoformat() if user.updated_at else None,
    }


@router.post("/{user_id}/follow")
async def follow(
    user_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    target = await get_user_by_id(db, user_id)
    if not target:
        raise NotFoundException("资源不存在")
    if target.allow_follow == 0:
        raise AuthorizationException("无权限访问")
    ok = await follow_user(db, current_user.id, user_id)
    return success_response(data={"ok": ok, "following": True}, message="关注成功")


@router.delete("/{user_id}/follow")
async def unfollow(
    user_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    ok = await unfollow_user(db, current_user.id, user_id)
    return success_response(data={"ok": ok, "following": False}, message="取消关注成功")


@router.get("/me/followers")
async def my_followers(
    limit: int = Query(20, ge=1, le=50),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    users, total = await get_followers(db, current_user.id, limit=limit, offset=offset)
    data = {
        "items": [_user_to_response(u) for u in users],
        "total": total,
    }
    return success_response(data=data, message="获取粉丝列表成功")


@router.get("/me/feed")
async def my_feed(
    limit: int = Query(20, ge=1, le=50),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """关注流：当前用户关注的人发布的帖子（与 PRD 5.1.7、技术方案 3.3.2 一致）。"""
    total = await count_following_posts(db, current_user.id)
    posts = await list_following_posts(db, current_user.id, limit=limit, offset=offset)
    data = {
        "items": [PostResponse.model_validate(p).model_dump(mode='json') for p in posts],
        "total": total,
    }
    return success_response(data=data, message="获取关注流成功")


@router.get("/me/following")
async def my_following(
    limit: int = Query(20, ge=1, le=50),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    users, total = await get_following(db, current_user.id, limit=limit, offset=offset)
    data = {
        "items": [_user_to_response(u) for u in users],
        "total": total,
    }
    return success_response(data=data, message="获取关注列表成功")


@router.get("/{user_id}/followers")
async def user_followers(
    user_id: str,
    limit: int = Query(20, ge=1, le=50),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
):
    if not await get_user_by_id(db, user_id):
        raise NotFoundException("资源不存在")
    users, total = await get_followers(db, user_id, limit=limit, offset=offset)
    data = {
        "items": [_user_to_response(u) for u in users],
        "total": total,
    }
    return success_response(data=data, message="获取粉丝列表成功")


@router.get("/{user_id}/following")
async def user_following(
    user_id: str,
    limit: int = Query(20, ge=1, le=50),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
):
    if not await get_user_by_id(db, user_id):
        raise NotFoundException("资源不存在")
    users, total = await get_following(db, user_id, limit=limit, offset=offset)
    data = {
        "items": [_user_to_response(u) for u in users],
        "total": total,
    }
    return success_response(data=data, message="获取关注列表成功")

File: miniapp/src/api/follow.ts
/**
 * 关注、关注流接口 - 与 backend app.api.v1.follow 一致
 */
import request from '@/utils/request'

const PREFIX = '/api/v1/user'

/** 关注/取关响应 */
export interface FollowActionRes {
  ok: boolean
  following: boolean
}

/** 用户项（粉丝/关注列表） */
export interface UserItem {
  id: string
  anonymous_name: string
  avatar: string | null
  bio: string | null
  follower_count: number
  following_count: number
  post_count: number
  allow_follow: number
  allow_comment: number
  status: number
  created_at: string | null
  updated_at: string | null
}

/** 列表响应 */
export interface FollowListRes {
  items: UserItem[]
  total: number
}

/** 关注流帖子项 */
export interface FeedPost {
  id: string
  user_id: string
  anonymous_name: string
  content: string
  images: string[] | null
  tags: string[] | null
  type: string
  salary_range: string | null
  industry: string | null
  city: string | null
  view_count: number
  like_count: number
  comment_count: number
  status: string
  risk_status: string
  risk_score: number | null
  risk_reason: string | null
  created_at: string
  updated_at: string
}

/** 关注流响应 */
export interface FeedRes {
  items: FeedPost[]
  total: number
}

/** 关注用户 */
export function followUser(userId: string) {
  return request<FollowActionRes>({
    url: `${PREFIX}/${userId}/follow`,
    method: 'POST',
  })
}

/** 取关用户 */
export function unfollowUser(userId: string) {
  return request<FollowActionRes>({
    url: `${PREFIX}/${userId}/follow`,
    method: 'DELETE',
  })
}

function queryString(params: Record<string, number | string>): string {
  const queryParts = Object.entries(params).map(([k, v]) => `${k}=${encodeURIComponent(String(v))}`)
  const q = queryParts.join('&')
  return q ? `?${q}` : ''
}

/** 我的关注流（我关注的人发的帖子） */
export function getMyFeed(params?: { limit?: number; offset?: number }) {
  const limit = params?.limit ?? 20
  const offset = params?.offset ?? 0
  return request<FeedRes>({
    url: `${PREFIX}/me/feed${queryString({ limit, offset })}`,
    method: 'GET',
  })
}

/** 我的粉丝列表 */
export function getMyFollowers(params?: { limit?: number; offset?: number }) {
  const limit = params?.limit ?? 20
  const offset = params?.offset ?? 0
  return request<FollowListRes>({
    url: `${PREFIX}/me/followers${queryString({ limit, offset })}`,
    method: 'GET',
  })
}

/** 我的关注列表 */
export function getMyFollowing(params?: { limit?: number; offset?: number }) {
  const limit = params?.limit ?? 20
  const offset = params?.offset ?? 0
  return request<FollowListRes>({
    url: `${PREFIX}/me/following${queryString({ limit, offset })}`,
    method: 'GET',
  })
}

/** 某用户的粉丝列表（匿名主页用） */
export function getUserFollowers(userId: string, params?: { limit?: number; offset?: number }) {
  const limit = params?.limit ?? 20
  const offset = params?.offset ?? 0
  return request<FollowListRes>({
    url: `${PREFIX}/${userId}/followers${queryString({ limit, offset })}`,
    method: 'GET',
  })
}

/** 某用户的关注列表 */
export function getUserFollowing(userId: string, params?: { limit?: number; offset?: number }) {
  const limit = params?.limit ?? 20
  const offset = params?.offset ?? 0
  return request<FollowListRes>({
    url: `${PREFIX}/${userId}/following${queryString({ limit, offset })}`,
    method: 'GET',
  })
}

/** 检查关注状态响应 */
export interface CheckFollowStatusRes {
  is_following: boolean
}

/** 检查是否关注某用户 */
export function checkFollowStatus(userId: string) {
  return request<CheckFollowStatusRes>({
    url: `${PREFIX}/${userId}/follow-status`,
    method: 'GET',
  })
}

/** 批量检查关注状态响应 */
export type BatchFollowStatusRes = Record<string, boolean>

/** 批量检查关注状态 */
export function checkBatchFollowStatus(userIds: string[]): Promise<BatchFollowStatusRes> {
  return request<BatchFollowStatusRes>({
    url: '/api/v1/follows/status',
    method: 'POST',
    data: { user_ids: userIds },
  })
}
